### PR TITLE
started to add the possibility to keep a model on hold

### LINKED ISSUE #271: hold models in ensemble configuration for multilevel optimization
--------
Issue Description
--------
##### What did you expect to see happen?
When a multiple layer optimization is performed, I would like to have the possibility to hold the outcome of a model (in the ensemble model configuration). This can be beneficial for intrinsically stochastic models.

##### What did you see instead?
not implemented

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [ ] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #271

##### What are the significant changes in functionality due to this change request?
It adds the possibility to hold models in multilevel optimization (the coding should be revised since it is very dirty but the capability is needed by the NHES project)

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
I still don't quite understand what "holding" part of the outspace means, and why it's useful.  Could you help me understand it a little?

**Comment by alfoa** [comment]
@AaronEpiney

**Comment by alfoa** [comment]
I added an issue

File: framework/DataObjects/Data.py [hunk 1]
```diff
@@ -1166,3 +1166,31 @@ def __getMetadataType(self,name,value):
       valueType = None if utils.checkTypeRecursively(value) not in ['str','unicode','bytes'] else object
       self._dataParameters['typeMetadata'][name] = valueType
     return valueType
+
+  #def getRealizationGivenEvaluationID(self, evaluationID, variable=None):
```
Comment (PaulTalbot-INL) [inline framework/DataObjects/Data.py]: Is there a reason we're keeping this code?

File: framework/Models/EnsembleModel.py [hunk 2]
```diff
@@ -445,6 +446,7 @@ def collectOutput(self,finishedJob,output):
       if jobIndex is not None:
         for optionalModelOutput in self.modelsDictionary[modelIn]['OutputObject']:
           self.modelsDictionary[modelIn]['Instance'].collectOutput(finishedJob,optionalModelOutput,options={'exportDict':copy.copy(optionalOutputs[modelIn])})
+    #self.tempOutputs['forHold'][finishedJob.identifier] = exportDictTargetEvaluation
```
Comment (PaulTalbot-INL) [inline framework/Models/EnsembleModel.py]: commented code; please add commentary if it's being kept on purpose.

File: tests/framework/Optimizers/MTMLwithEnsembleAndModelHolding/amplifier.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# takes input parameters ans and z
+# returns value in "objAns" amplifying the self.ans times (self.z+1.0)
+
+def run(self,Inputs):
```
Comment (PaulTalbot-INL) [inline tests/framework/Optimizers/MTMLwithEnsembleAndModelHolding/amplifier.py]: docstrings

File: framework/DataObjects/Data.py (hunk 1 above)
Comment (alfoa) [inline framework/DataObjects/Data.py]: removed

File: framework/Models/EnsembleModel.py (hunk 2 above)
Comment (alfoa) [inline framework/Models/EnsembleModel.py]: removed

File: tests/framework/Optimizers/MTMLwithEnsembleAndModelHolding/amplifier.py (hunk 3 above)
Comment (alfoa) [inline tests/framework/Optimizers/MTMLwithEnsembleAndModelHolding/amplifier.py]: added

**Comment by alfoa** [comment]
Addressed @PaulTalbot-INL 's comments

**Comment by PaulTalbot-INL** [comment]
Did the changes get pushed? I don't see the commit.

**Comment by alfoa** [comment]
Sorry @PaulTalbot-INL... I forgot to push.

Now you should see the modifications.

**Pull request merged**